core/arch/riscv/mm/core_mmu_arch.c

```diff
@@ -0,0 +1,721 @@
+						unsigned int idx)
+{
+	return &pgt->entries[idx & RISCV_MMU_VPN_MASK];
+}
+
+static void core_mmu_entry_set(struct mmu_pte *pte, uint64_t val)
+{
+	pte->entry = val;
+}
+
+static uint64_t core_mmu_entry_get(struct mmu_pte *pte)
+{
+	return pte->entry;
+}
+
+static bool core_mmu_entry_is_valid(struct mmu_pte *pte)
+{
+	return pte->entry & PTE_V;
+}
+
+static bool core_mmu_entry_is_invalid(struct mmu_pte *pte)
+{
+	return !core_mmu_entry_is_valid(pte);
+}
+
+static bool core_mmu_entry_is_leaf(struct mmu_pte *pte)
+{
+	/* A leaf has one or more RWX bits set */
+	return pte->entry & (PTE_R | PTE_W | PTE_X);
+}
+
+static bool __maybe_unused core_mmu_entry_is_branch(struct mmu_pte *pte)
+{
+	return !core_mmu_entry_is_leaf(pte);
+}
+
+static unsigned long core_mmu_pte_create(unsigned long ppn, uint8_t perm)
+{
+	return SHIFT_U64(ppn, PTE_PPN_SHIFT) | PTE_V | perm;
+}
+
+static unsigned long core_mmu_ptp_create(unsigned long ppn)
+{
+	/* set perms to 0 since core_mmu_pte_create() already adds PTE_V */
+	return core_mmu_pte_create(ppn, 0);
+}
+
+static unsigned long core_mmu_pte_ppn(struct mmu_pte *pte)
+{
+	return pte->entry >> PTE_PPN_SHIFT;
+}
+
+static unsigned long pa_to_ppn(paddr_t pa)
+{
+	return pa >> RISCV_PGSHIFT;
+}
+
+static paddr_t pte_to_pa(struct mmu_pte *pte)
+{
+	return SHIFT_U64(core_mmu_pte_ppn(pte), RISCV_PGSHIFT);
+}
+
+static unsigned long core_mmu_pgt_to_satp(unsigned long asid,
+					  struct mmu_pgt *pgt)
+{
+	unsigned long satp = 0;
+	unsigned long pgt_ppn = virt_to_phys(pgt) >> RISCV_PGSHIFT;
+
+	assert(asid & g_asid == asid);
+	satp |= SHIFT_U64(asid, RISCV_SATP_ASID_SHIFT);
+	satp |= SHIFT_U64(RISCV_SATP_MODE, RISCV_SATP_MODE_SHIFT);
+	satp |= pgt_ppn;
+
+	return satp;
+}
+
+static unsigned long pte_to_mattr(unsigned level __maybe_unused,
+				  struct mmu_pte *pte)
+{
+	unsigned long mattr = TEE_MATTR_SECURE;
+	unsigned long entry = core_mmu_entry_get(pte);
+
+	if (entry & PTE_V) {
+		if (!(entry & (PTE_R | PTE_W | PTE_X)))
+			return TEE_MATTR_TABLE;
+
+		mattr |=  TEE_MATTR_VALID_BLOCK;
+	}
+
+	if (entry & PTE_U) {
+		if (entry & PTE_R)
+			mattr |= TEE_MATTR_UR;
+		if (entry & PTE_W)
+			mattr |= TEE_MATTR_UW;
+		if (entry & PTE_X)
+			mattr |= TEE_MATTR_UX;
+	} else {
+		if (entry & PTE_R)
+			mattr |= TEE_MATTR_PR;
+		if (entry & PTE_W)
+			mattr |= TEE_MATTR_PW;
+		if (entry & PTE_X)
+			mattr |= TEE_MATTR_PX;
+	}
+
+	if (entry & PTE_G)
+		mattr |= TEE_MATTR_GLOBAL;
+
+	return mattr;
+}
+
+static uint8_t mattr_to_perms(unsigned level __maybe_unused,
+			      uint32_t attr)
+{
+	unsigned long perms = 0;
+
+	if (attr & TEE_MATTR_TABLE)
+		return PTE_V;
+
+	if (attr & TEE_MATTR_VALID_BLOCK)
+		perms |= PTE_V;
+
+	if (attr & TEE_MATTR_UR)
+		perms |= PTE_R | PTE_U;
+	if (attr & TEE_MATTR_UW)
+		perms |= PTE_W | PTE_U;
+	if (attr & TEE_MATTR_UX)
+		perms |= PTE_X | PTE_U;
+
+	if (attr & TEE_MATTR_PR)
+		perms |= PTE_R;
+	if (attr & TEE_MATTR_PW)
+		perms |= PTE_W;
```

Comment:
What happens if `PTE_R` isn't set, can you get write-only memory on riscv?

Reply:
According to RISC-V privileged spec, page table entry with WO or WX permission bits are reserved for future use.
The two patterns don't have special meaning at this moment.

@maroueneboubakri , how about appending `PTE_R` even if `attr` only has write or execute permissions?